File: apps/client/local_plugins/core/knowledge_native/ios/Classes/KnowledgeCore.swift
import Foundation

// 知识库算法核心（Swift）——镜像 KnowledgeCore.kt / 已验证的 Dart 核心。
// 供纯原生 agents_server 后台/唤醒直调；默认零外部依赖（内存索引 + 特征哈希嵌入）。

// MARK: - 分词

/// ASCII 词 + 中文单字/相邻双字(bigram)。与 Kotlin/Dart 一致。
func tokenize(_ input: String) -> [String] {
    var tokens: [String] = []
    var buf = ""
    var prevCjk: UnicodeScalar? = nil
    func flush() { if !buf.isEmpty { tokens.append(buf); buf = "" } }
    for scalar in input.lowercased().unicodeScalars {
        let cp = scalar.value
        let isAsciiAlnum = (cp >= 0x30 && cp <= 0x39) || (cp >= 0x61 && cp <= 0x7a)
        let isCjk = cp >= 0x4E00 && cp <= 0x9FFF
        if isAsciiAlnum {
            buf.unicodeScalars.append(scalar)
            prevCjk = nil
        } else if isCjk {
            flush()
            tokens.append(String(scalar))
            if let p = prevCjk { tokens.append(String(p) + String(scalar)) }
            prevCjk = scalar
        } else {
            flush()
            prevCjk = nil
        }
    }
    flush()
    return tokens
}

/// 确定性 32-bit FNV-1a（按 UTF-16 code unit，跨运行稳定）。
func fnv1a(_ s: String) -> UInt64 {
    var h: UInt64 = 0x811c9dc5
    for u in s.utf16 {
        h ^= UInt64(u)
        h = (h &* 0x01000193) & 0xffffffff
    }
    return h
}

// MARK: - 模型

struct ChunkLoc { let start: Int; let end: Int; let heading: String? }

struct KbChunk {
    let id: String
    let documentId: String
    let userId: String
    let ordinal: Int
    let text: String
    let loc: ChunkLoc
}

struct KbDocument {
    let id: String
    let sourceId: String
    let userId: String
    let title: String?
    let createdAt: Int64
}

struct ScoredChunk { let chunk: KbChunk; let score: Double; let documentTitle: String? }

// MARK: - 端口

protocol EmbeddingProvider {
    var model: String { get }
    var dim: Int { get }
    func embed(_ text: String) -> [Double]
}
extension EmbeddingProvider {
    func embedBatch(_ texts: [String]) -> [[Double]] { texts.map { embed($0) } }
}

struct VectorRecord { let chunkId: String; let documentId: String; let userId: String; let vector: [Double] }

protocol VectorStore {
    func upsert(_ records: [VectorRecord])
    /// 返回 (chunkId, 余弦相似度) 降序 topK。
    func search(_ query: [Double], k: Int, userId: String?) -> [(String, Double)]
    func deleteByDocument(_ documentId: String)
    func clear()
    func count() -> Int
}

protocol KnowledgeStore {
    func putDocument(_ doc: KbDocument)
    func putChunks(_ chunks: [KbChunk])
    func getChunk(_ id: String) -> KbChunk?
    func getDocument(_ id: String) -> KbDocument?
    func allChunks(userId: String?) -> [KbChunk]
    func deleteDocument(_ documentId: String)
    func clearIndex()
}

// MARK: - 端上哈希嵌入（默认/离线）

/// 平台层可用 ONNX 本地模型（bge-small-zh/e5-small）实现同一协议替换。
final class HashingEmbedding: EmbeddingProvider {
    let dim: Int
    init(dim: Int = 256) { self.dim = dim }
    var model: String { "hashing-v1-\(dim)" }
    func embed(_ text: String) -> [Double] {
        var v = [Double](repeating: 0, count: dim)
        for tok in tokenize(text) {
            let h = fnv1a(tok)
            let idx = Int(h % UInt64(dim))
            let sign = (fnv1a("#" + tok) & 1) == 0 ? 1.0 : -1.0
            v[idx] += sign
        }
        var norm = 0.0
        for x in v { norm += x * x }
        norm = norm.squareRoot()
        if norm > 0 { for i in 0..<dim { v[i] /= norm } }
        return v
    }
}

func cosine(_ a: [Double], _ b: [Double]) -> Double {
    let n = min(a.count, b.count)
    var dot = 0.0, na = 0.0, nb = 0.0
    for i in 0..<n { dot += a[i] * b[i]; na += a[i] * a[i]; nb += b[i] * b[i] }
    if na == 0 || nb == 0 { return 0 }
    return dot / (na.squareRoot() * nb.squareRoot())
}

// MARK: - 切块

final class Chunker {
    let maxChars: Int
    let overlap: Int
    init(maxChars: Int = 800, overlap: Int = 120) { self.maxChars = maxChars; self.overlap = overlap }

    func chunk(_ markdown: String) -> [(String, ChunkLoc)] {
        var out: [(String, ChunkLoc)] = []
        let paras = splitParagraphs(markdown)
        var buf = ""
        var bufStart = 0
        var heading: String? = nil
        func flush(_ end: Int) {
            let t = buf.trimmingCharacters(in: .whitespacesAndNewlines)
            if !t.isEmpty { out.append((t, ChunkLoc(start: bufStart, end: end, heading: heading))) }
            buf = ""
        }
        for (text, start) in paras {
            if isHeading(text) { heading = stripHeading(text) }
            if !buf.isEmpty && buf.count + text.count > maxChars {
                flush(start)
                let prev = out.last?.0 ?? ""
                let tail = prev.count > overlap ? String(prev.suffix(overlap)) : prev
                buf += tail
                bufStart = start
            }
            if buf.isEmpty { bufStart = start } else { buf += "\n\n" }
            buf += text
        }
        flush(markdown.count)
        return out
    }

    private func isHeading(_ s: String) -> Bool {
        var hashes = 0
        for ch in s {
            if ch == "#" { hashes += 1 }
            else { return hashes >= 1 && hashes <= 6 && (ch == " " || ch == "\t") }
        }
        return false
    }

    private func stripHeading(_ s: String) -> String {
        var i = s.startIndex
        while i < s.endIndex && s[i] == "#" { i = s.index(after: i) }
        return String(s[i...]).trimmingCharacters(in: .whitespaces)
    }

    private func splitParagraphs(_ md: String) -> [(String, Int)] {
        var res: [(String, Int)] = []
        var cursor = md.startIndex
        for raw in md.components(separatedBy: "\n\n") {
            guard !raw.isEmpty,
                  let range = md.range(of: raw, range: cursor..<md.endIndex) else { continue }
            let start = md.distance(from: md.startIndex, to: range.lowerBound)
            let t = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            if !t.isEmpty { res.append((t, start)) }
            cursor = range.upperBound
        }
        return res
    }
}

// MARK: - 内存存储（默认实现）

final class InMemoryVectorStore: VectorStore {
    private var byChunk: [String: VectorRecord] = [:]
    func upsert(_ records: [VectorRecord]) { for r in records { byChunk[r.chunkId] = r } }
    func search(_ query: [Double], k: Int, userId: String?) -> [(String, Double)] {
        byChunk.values
            .filter { userId == nil || $0.userId == userId }
            .map { ($0.chunkId, cosine(query, $0.vector)) }
            .sorted { $0.1 > $1.1 }
            .prefix(k)
            .map { $0 }
    }
    func deleteByDocument(_ documentId: String) {
        byChunk = byChunk.filter { $0.value.documentId != documentId }
    }
    func clear() { byChunk.removeAll() }
    func count() -> Int { byChunk.count }
}

final class InMemoryKnowledgeStore: KnowledgeStore {
    private var docs: [String: KbDocument] = [:]
    private var chunks: [String: KbChunk] = [:]
    func putDocument(_ doc: KbDocument) { docs[doc.id] = doc }
    func putChunks(_ chunkList: [KbChunk]) { for c in chunkList { chunks[c.id] = c } }
    func getChunk(_ id: String) -> KbChunk? { chunks[id] }
    func getDocument(_ id: String) -> KbDocument? { docs[id] }
    func allChunks(userId: String?) -> [KbChunk] {
        chunks.values.filter { userId == nil || $0.userId == userId }
    }
    func deleteDocument(_ documentId: String) {
        docs[documentId] = nil
        chunks = chunks.filter { $0.value.documentId != documentId }
    }
    func clearIndex() { docs.removeAll(); chunks.removeAll() }
}

// MARK: - 混合检索（向量 + 关键词 RRF）

final class Retriever {
    private let vectors: VectorStore
    private let store: KnowledgeStore
    private let embedder: EmbeddingProvider
    private let rrfK: Int
    init(vectors: VectorStore, store: KnowledgeStore, embedder: EmbeddingProvider, rrfK: Int = 60) {
        self.vectors = vectors; self.store = store; self.embedder = embedder; self.rrfK = rrfK
    }

    func retrieve(_ query: String, k: Int = 6, userId: String? = nil) -> [ScoredChunk] {
        let qvec = embedder.embed(query)
        let vHits = vectors.search(qvec, k: k * 3, userId: userId).map { $0.0 }
        let kHits = keyword(query, k: k * 3, userId: userId)

        var fused: [String: Double] = [:]
        func fuse(_ ids: [String]) {
            for (i, id) in ids.enumerated() { fused[id, default: 0] += 1.0 / Double(rrfK + i + 1) }
        }
        fuse(vHits); fuse(kHits)

        return fused.sorted { $0.value > $1.value }.prefix(k).compactMap { e in
            guard let c = store.getChunk(e.key) else { return nil }
            let doc = store.getDocument(c.documentId)
            return ScoredChunk(chunk: c, score: e.value, documentTitle: doc?.title)
        }
    }

    private func keyword(_ query: String, k: Int, userId: String?) -> [String] {
        let q = Set(tokenize(query))
        if q.isEmpty { return [] }
        return store.allChunks(userId: userId).compactMap { c -> (String, Double)? in
            let toks = tokenize(c.text)
            if toks.isEmpty { return nil }
            let hit = toks.filter { q.contains($0) }.count
            return hit > 0 ? (c.id, Double(hit) / Double(toks.count)) : nil
        }
        .sorted { $0.1 > $1.1 }
        .prefix(k)
        .map { $0.0 }
    }
}

File: apps/client/local_plugins/core/knowledge_native/android/src/main/kotlin/ai/unihelper/knowledge/KnowledgeCore.kt
package ai.unihelper.knowledge

import kotlin.math.sqrt

/**
 * 知识库算法核心（Kotlin）——忠实移植已验证的 Dart `core/knowledge`。
 *
 * 供纯原生 `agents_server` 在后台/设备唤醒场景直接调用（不经 Flutter）。
 * 默认实现零外部依赖（内存向量 + 特征哈希嵌入），可先跑通；
 * sqlite-vec / ONNX 端上嵌入以同名端口替换（见 [VectorStore] / [EmbeddingProvider]）。
 */

// ─────────────────────────── 分词 ───────────────────────────

/** ASCII 词 + 中文单字/相邻双字(bigram)。与 Dart tokenizer 一致。 */
fun tokenize(input: String): List<String> {
    val tokens = ArrayList<String>()
    val buf = StringBuilder()
    var prevCjk: Char? = null
    fun flush() {
        if (buf.isNotEmpty()) {
            tokens.add(buf.toString()); buf.setLength(0)
        }
    }
    for (c in input.lowercase()) {
        val cp = c.code
        val isAsciiAlnum = (cp in 0x30..0x39) || (cp in 0x61..0x7a)
        val isCjk = cp in 0x4E00..0x9FFF
        when {
            isAsciiAlnum -> { buf.append(c); prevCjk = null }
            isCjk -> {
                flush()
                tokens.add(c.toString())
                val p = prevCjk
                if (p != null) tokens.add("$p$c")
                prevCjk = c
            }
            else -> { flush(); prevCjk = null }
        }
    }
    flush()
    return tokens
}

/** 确定性 32-bit FNV-1a（跨运行稳定）。 */
fun fnv1a(s: String): Long {
    var h = 0x811c9dc5L
    for (ch in s) {
        h = h xor ch.code.toLong()
        h = (h * 0x01000193L) and 0xffffffffL
    }
    return h
}

// ─────────────────────────── 模型 ───────────────────────────

data class ChunkLoc(val start: Int, val end: Int, val heading: String?)

data class KbChunk(
    val id: String,
    val documentId: String,
    val userId: String,
    val ordinal: Int,
    val text: String,
    val loc: ChunkLoc,
)

data class KbDocument(
    val id: String,
    val sourceId: String,
    val userId: String,
    val title: String?,
    val createdAt: Long,
)

data class ScoredChunk(val chunk: KbChunk, val score: Double, val documentTitle: String?)

// ─────────────────────────── 端口 ───────────────────────────

interface EmbeddingProvider {
    val model: String
    val dim: Int
    fun embed(text: String): DoubleArray
    fun embedBatch(texts: List<String>): List<DoubleArray> = texts.map { embed(it) }
}

data class VectorRecord(
    val chunkId: String,
    val documentId: String,
    val userId: String,
    val vector: DoubleArray,
)

interface VectorStore {
    fun upsert(records: List<VectorRecord>)
    /** 返回 (chunkId, 余弦相似度) 降序 topK。 */
    fun search(query: DoubleArray, k: Int = 8, userId: String? = null): List<Pair<String, Double>>
    fun deleteByDocument(documentId: String)
    fun clear()
    fun count(): Int
}

interface KnowledgeStore {
    fun putDocument(doc: KbDocument)
    fun putChunks(chunks: List<KbChunk>)
    fun getChunk(id: String): KbChunk?
    fun getDocument(id: String): KbDocument?
    fun allChunks(userId: String? = null): List<KbChunk>
    fun deleteDocument(documentId: String)
    fun clearIndex()
}

// ─────────────────────── 端上哈希嵌入（默认/离线） ───────────────────────

/** 平台层可用 ONNX 本地模型（bge-small-zh/e5-small）实现同一端口替换。 */
class HashingEmbedding(override val dim: Int = 256) : EmbeddingProvider {
    override val model: String get() = "hashing-v1-$dim"
    override fun embed(text: String): DoubleArray {
        val v = DoubleArray(dim)
        for (tok in tokenize(text)) {
            val h = fnv1a(tok)
            val idx = (h % dim).toInt()
            val sign = if ((fnv1a("#$tok") and 1L) == 0L) 1.0 else -1.0
            v[idx] += sign
        }
        var norm = 0.0
        for (x in v) norm += x * x
        norm = sqrt(norm)
        if (norm > 0) for (i in v.indices) v[i] /= norm
        return v
    }
}

// ─────────────────────────── 切块 ───────────────────────────

class Chunker(private val maxChars: Int = 800, private val overlap: Int = 120) {
    private val headingRe = Regex("^#{1,6}\\s")
    private val leadHashRe = Regex("^#+\\s*")
    private val paraSplit = Regex("\\n\\s*\\n")

    fun chunk(markdown: String): List<Pair<String, ChunkLoc>> {
        val out = ArrayList<Pair<String, ChunkLoc>>()
        val paras = splitParagraphs(markdown)
        val buf = StringBuilder()
        var bufStart = 0
        var heading: String? = null
        fun flush(end: Int) {
            val t = buf.toString().trim()
            if (t.isNotEmpty()) out.add(t to ChunkLoc(bufStart, end, heading))
            buf.setLength(0)
        }
        for ((text, start) in paras) {
            if (headingRe.containsMatchIn(text)) {
                heading = text.replaceFirst(leadHashRe, "").trim()
            }
            if (buf.isNotEmpty() && buf.length + text.length > maxChars) {
                flush(start)
                val prev = if (out.isNotEmpty()) out.last().first else ""
                val tail = if (prev.length > overlap) prev.substring(prev.length - overlap) else prev
                buf.append(tail)
                bufStart = start
            }
            if (buf.isEmpty()) bufStart = start else buf.append("\n\n")
            buf.append(text)
        }
        flush(markdown.length)
        return out
    }

    private fun splitParagraphs(md: String): List<Pair<String, Int>> {
        val res = ArrayList<Pair<String, Int>>()
        var cursor = 0
        for (raw in md.split(paraSplit)) {
            val idx = md.indexOf(raw, cursor)
            val start = if (idx == -1) cursor else idx
            val t = raw.trim()
            if (t.isNotEmpty()) res.add(t to start)
            cursor = start + raw.length
        }
        return res
    }
}

// ─────────────────────── 内存存储（默认实现） ───────────────────────

class InMemoryVectorStore : VectorStore {
    private val byChunk = LinkedHashMap<String, VectorRecord>()
    override fun upsert(records: List<VectorRecord>) {
        for (r in records) byChunk[r.chunkId] = r
    }
    override fun search(query: DoubleArray, k: Int, userId: String?): List<Pair<String, Double>> =
        byChunk.values
            .filter { userId == null || it.userId == userId }
            .map { it.chunkId to cosine(query, it.vector) }
            .sortedByDescending { it.second }
            .take(k)
    override fun deleteByDocument(documentId: String) {
        byChunk.values.removeAll { it.documentId == documentId }
    }
    override fun clear() = byChunk.clear()
    override fun count(): Int = byChunk.size
}

class InMemoryKnowledgeStore : KnowledgeStore {
    private val docs = LinkedHashMap<String, KbDocument>()
    private val chunks = LinkedHashMap<String, KbChunk>()
    override fun putDocument(doc: KbDocument) { docs[doc.id] = doc }
    override fun putChunks(chunkList: List<KbChunk>) { for (c in chunkList) chunks[c.id] = c }
    override fun getChunk(id: String): KbChunk? = chunks[id]
    override fun getDocument(id: String): KbDocument? = docs[id]
    override fun allChunks(userId: String?): List<KbChunk> =
        chunks.values.filter { userId == null || it.userId == userId }
    override fun deleteDocument(documentId: String) {
        docs.remove(documentId)
        chunks.values.removeAll { it.documentId == documentId }
    }
    override fun clearIndex() { docs.clear(); chunks.clear() }
}

internal fun cosine(a: DoubleArray, b: DoubleArray): Double {
    val n = minOf(a.size, b.size)
    var dot = 0.0; var na = 0.0; var nb = 0.0
    for (i in 0 until n) { dot += a[i] * b[i]; na += a[i] * a[i]; nb += b[i] * b[i] }
    if (na == 0.0 || nb == 0.0) return 0.0
    return dot / (sqrt(na) * sqrt(nb))
}

// ─────────────────────── 混合检索（向量 + 关键词 RRF） ───────────────────────

class Retriever(
    private val vectors: VectorStore,
    private val store: KnowledgeStore,
    private val embedder: EmbeddingProvider,
    private val rrfK: Int = 60,
) {
    fun retrieve(query: String, k: Int = 6, userId: String? = null): List<ScoredChunk> {
        val qvec = embedder.embed(query)
        val vHits = vectors.search(qvec, k * 3, userId).map { it.first }
        val kHits = keyword(query, k * 3, userId)

        val fused = LinkedHashMap<String, Double>()
        fun fuse(ids: List<String>) {
            for (i in ids.indices) fused[ids[i]] = (fused[ids[i]] ?: 0.0) + 1.0 / (rrfK + i + 1)
        }
        fuse(vHits); fuse(kHits)

        return fused.entries
            .sortedByDescending { it.value }
            .take(k)
            .mapNotNull { e ->
                val c = store.getChunk(e.key) ?: return@mapNotNull null
                val doc = store.getDocument(c.documentId)
                ScoredChunk(c, e.value, doc?.title)
            }
    }

    private fun keyword(query: String, k: Int, userId: String?): List<String> {
        val qTokens = tokenize(query).toHashSet()
        if (qTokens.isEmpty()) return emptyList()
        return store.allChunks(userId)
            .mapNotNull { c ->
                val toks = tokenize(c.text)
                if (toks.isEmpty()) return@mapNotNull null
                val hit = toks.count { it in qTokens }
                if (hit > 0) c.id to hit.toDouble() / toks.size else null
            }
            .sortedByDescending { it.second }
            .take(k)
            .map { it.first }
    }
}
